book/src/dev/rfcs/drafts/0005-treestate.md

```diff
@@ -179,9 +179,18 @@ time).
 
 IMPORTANT: we need to save the incremental merkle tree / serialized nodes for:
 
-Sapling tip block
-~all Sprout blocks
-chains we're tracking in memory within the reorg limit, for both
+Orchard tip: 
+- When finalizing state, for Orchard, the tip of the chain (latest finalized block) saves a serialization of the Orchard Note Commitment Tree
+- When non-finalized state is being updated, each non-finalized chain gets its own copy of the Orchard Note Commitment Tree, cloned from the finalized block note commitment tree of its parent block, and then subsequently updated/extended as new Orchard note commitments are discovered when processing blocks.
+
+Sapling tip: 
+- When finalizing state, for Sapling, the tip of the chain (latest finalized block) saves a serialization of the Sapling Note Commitment Tree
+- When non-finalized state is being updated, each non-finalized chain gets its own copy of the Sapling Note Commitment Tree, cloned from the finalized block note commitment tree of its parent block, and then subsequently updated/extended as new Sapling note commitments are discovered when processing blocks.
+
+All Sprout blocks:
+- Every finalized block needs its own copy of the Sprout note commitment tree accurate to that block (😿)
+- When non-finalized state is being updated, each non-finalized chain gets its own copy of the Note Commitment Tree, cloned from the finalized block note commitment tree of its parent block, and then subsequently updated/extended as new Sprout note commitments are discovered when processing blocks.
+
 We can't just compute a fresh tree with just the note commitments within a block, we are adding them to the tree referenced by the anchor, but we cannot update that tree with just the anchor, we need the 'frontier' nodes and leaves of the incremental merkle tree.
 
 # Reference-level explanation
```

Comment:
Unanswered questions about the finalized state:

1. What keys do we need to be able to look up quickly? For transaction validation? For treestate updates?

This changes how we index treestates in the state.

The state already has sprout and sapling anchors, and sprout, sapling, and orchard nullifiers.

2. Are there any rules about which treestates can be used for which transactions?

This might change how we maintain (Sprout?) treestates in the non-finalized state.

For example, can Sprout, Sapling, and Orchard transactions use the previous block's treestate?
(There are a lot of references to "finalized" treestates in the rest of the RFC, but I'm not sure if they are the same as the "finalized" state.)

The rest of the RFC mentions that Sprout transactions can use the treestate of any previous transaction in the block.

3. Why do we need to store a separate Sprout treestate for every block? Can we skip blocks without Sprout transactions? Are there parts of the treestate data that don't change?

This impacts the indexes, storage, and lookups for those treestates.

4. What are the details of a Sprout treestate update?

This might change how we maintain (Sprout?) treestates in the non-finalized or finalized state.

https://github.com/ZcashFoundation/zebra/blob/treestate-rfc-updates/book/src/dev/rfcs/0005-state-updates.md#rocksdb-data-structures

Reply:
@dconnolly do you want to schedule a design meeting to go over these questions?

Reply:
@teor2345 I'm tracking these questions as part of https://github.com/ZcashFoundation/zebra/issues/1320

Reply:
@teor2345  Let me know if you think we should address these questions as part of this PR or as part of #1320 tracking

Reply:
I think these questions are being addressed by the design in https://github.com/ZcashFoundation/zebra/issues/2425